third_party/nvfuser/csrc/ir_nodes.cpp

```diff
@@ -2996,4 +3074,209 @@ c10::optional<ParallelType> NamedScalar::getParallelIndex() const {
+}
+
+std::vector<int> PadOp::getPaddedAxes() const {
+  auto num_dims = out()->as<TensorView>()->getRootDomain().size();
+  auto num_padded_axes = attributes_.size() / 2;
+  std::vector<int> padded_axes(num_padded_axes);
+  std::iota(padded_axes.begin(), padded_axes.end(), num_dims - num_padded_axes);
+  return padded_axes;
+}
+
+std::vector<Val*> PadOp::getPadWidths() const {
+  std::vector<Val*> pad_widths;
+  std::transform(
+      attributes_.begin(),
+      attributes_.end(),
+      std::back_inserter(pad_widths),
+      [](Statement* attr) { return attr->as<Val>(); });
+  return pad_widths;
+}
+
+std::pair<Val*, Val*> PadOp::getPadWidths(int axis) const {
+  auto num_dims = static_cast<int>(out()->as<TensorView>()->nDims());
+  auto num_padded_axes = static_cast<int>(attributes_.size() / 2);
+
+  if (axis < 0) {
+    axis += num_dims;
+  }
+
+  TORCH_CHECK(axis >= 0 && axis < num_dims, "Invalid axis: ", axis);
+
+  // Just return zero for non padded domains
+  if (axis < num_dims - num_padded_axes) {
+    return std::make_pair(container()->zeroVal(), container()->zeroVal());
+  }
+
+  auto pad_idx = axis - (num_dims - num_padded_axes);
+  TORCH_INTERNAL_ASSERT(pad_idx >= 0 && pad_idx < num_dims);
+
+  return std::make_pair(
+      attribute(pad_idx)->as<Val>(), attribute(pad_idx + 1)->as<Val>());
+}
+
+SliceOp::SliceOp(
+    IrBuilderPasskey passkey,
+    TensorView* out,
+    TensorView* inp,
+    const std::vector<Slice>& ranges)
+    : Expr(passkey) {
+  addOutput(out);
+  addInput(inp);
+  for (const auto& range : ranges) {
+    TORCH_INTERNAL_ASSERT(range.start != nullptr);
+    TORCH_INTERNAL_ASSERT(range.stop != nullptr);
+    TORCH_INTERNAL_ASSERT(range.step != nullptr);
+    addInput(range.start);
+    addInput(range.stop);
+    addInput(range.step);
+  }
+}
+
+NVFUSER_DEFINE_CLONE_AND_CREATE(SliceOp)
+
+std::string SliceOp::toString(int indent_size) const {
+  std::stringstream ss;
+  indent(ss, indent_size) << out()->toString() << "\n";
+  indent(ss, indent_size) << "   = slice( " << in()->toString() << ", {";
+  for (const auto& slice : getRanges()) {
+    ss << " {"
+       << toDelimitedString(std::vector<std::string>{
+              slice.start->toString(),
+              slice.stop->toString(),
+              slice.step->toString()})
+       << "}";
+  }
+  ss << " } )\n";
+  return ss.str();
+}
+
+std::string SliceOp::toInlineString(int indent_size) const {
+  TORCH_CHECK(false, "Tensor op can not be printed inline");
+}
+
+std::vector<Slice> SliceOp::getRanges() const {
+  const int input_offset = 1;
+  TORCH_INTERNAL_ASSERT((inputs().size() - input_offset) % 3 == 0);
+  auto ndims = (inputs().size() - input_offset) / 3;
+  std::vector<Slice> ranges(ndims);
+  for (const auto i : c10::irange(ndims)) {
+    ranges.at(i) = Slice{
+        .start = input(input_offset + i * 3)->as<Val>(),
+        .stop = input(input_offset + i * 3 + 1)->as<Val>(),
+        .step = input(input_offset + i * 3 + 2)->as<Val>()};
```

Comment:
I think inputs are already `Val`?
```suggestion
        .start = input(input_offset + i * 3),
        .stop = input(input_offset + i * 3 + 1),
        .step = input(input_offset + i * 3 + 2)};
```

Reply:
Ah, right. I guess this is because I originally set them as attributes. Will clean it up.